frontend/views/components/modal/Modal.vue

```diff
@@ -124,6 +135,9 @@ export default {
       // At the moment you can only replace a modal if it's the main one by design
       // Use direct children instead of sbp to wait for animation out
       this.$refs['content'].$children[0].close()
+    },
+    setModalQueries (componentName, queries) {
+      this.queries[componentName] = queries
```

Comment:
> In the modal, get the query at `created()` method. Make sure to call `SET_MODAL_QUERIES` with the query passed. This will guarantee that the query `username` gets removed when the modal is closed, even when the modal is opened through a direct link or page refresh.

This is rather cumbersome to have to always remember, and feels to me like a DRY violation, because the same redundant code must be copy/pasted and repeated in every modal. This is unclean.

When a modal is loaded, whether via `sbp('okTurtles.events/emit', OPEN_MODAL, 'RemoveMember', {username: 'bob' })`, or via a link, our modal system should automatically handle setting this variable. Surely there is a way for our modal system to listen to a page load / route change and set this query variable automatically?

Can any one of [these router guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#global-resolve-guards) be used to achieve this?

- via the `watch` key on `'$route'` above on line 43
- [Per-Route Guard](https://router.vuejs.org/guide/advanced/navigation-guards.html#per-route-guard) `beforeEnter`
- [In-Component Guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards): `beforeRouteEnter`, `beforeRouteUpdate`

Reply:
At the very least, each modal will have to record which query params it "owns" and thus deletes when closed, right? I mean, if you have `?foo=1&bar=2` but this modal only needs `foo`, when closed it should have `?bar=2`.

While `["foo"]` as a list of query params to delete could be obtained from the OPEN_MODAL call, when opening it from a URL there seems to be no way to distinguish this information, unless the modal itself were to provide it.

I'm not yet sure how/where to structure this, but these are at least my initial thoughts @taoeffect.

Reply:
@sdegutis Is it possible to structure the query/URL so that the parameters are associated with each modal in the URL itself?

Reply:
One of the simplest solutions is to rename each key like `${comp-name}-${prop-name}` e.g.:

* `?modal=UserSettingsModal&UserSettingsModal-action=my-account`
* `?modal=PaymentDetail&PaymentDetail-id=abc123`

It's more verbose, but gives us basic association. This would allow us to figure out more safely which query params to remove when the component closes, assuming we always retain a global namespace of component names. It would also protect against property name collisions, e.g. if the page you're on and the modal it opens both have a `filter` property for some reason.

Reply:
> > In the modal, get the query at `created()` method. Make sure to call `SET_MODAL_QUERIES` with the query passed. This will guarantee that the query `username` gets removed when the modal is closed, even when the modal is opened through a direct link or page refresh.
> 
> It seems rather cumbersome to have to always remember to do this, and feels to me like a DRY violation, because the same redundant code must be copy/pasted and repeated in every modal. Ideally it should be possible to use the given information to automatically do this.
> 
> When a modal is loaded, whether via `sbp('okTurtles.events/emit', OPEN_MODAL, 'RemoveMember', {username: 'bob' })`, or via a link, our modal system should automatically handle setting this variable. Surely there is a way for our modal system to listen to a page load / route change and set this query variable automatically?
> 
> Can any one of [these router guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#global-resolve-guards) be used to achieve this?
> 
> * via the `watch` key on `'$route'` above on line 43
> * [Per-Route Guard](https://router.vuejs.org/guide/advanced/navigation-guards.html#per-route-guard) `beforeEnter`
> * [In-Component Guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards): `beforeRouteEnter`, `beforeRouteUpdate`
> * [Global Before Guards](https://router.vuejs.org/guide/advanced/navigation-guards.html#global-before-guards): `beforeEach`

In order for the modal system to automatically "know" that certain query parameters belonged to the modal that's now exiting, some kind of association between modal component and its query parameters is needed:

1. When the modal component that sets queries mounts, and thus changes the routes, the modal system would need to know that this modal component is the one that set it, so that it can associate these queries with it; and when it unmounts, it needs to know which modal component is unmounting, so that it can remove them.

2. A similar solution is to put static data somewhere on the modal component itself, so that, when the modal component unmounts, it can look up which query fields to remove. This is basically what I had in mind in my first comment on this subthread.

Both of these solutions require that the router hooks have some way of identifying a specific *component* during route change events, which does not seem to be possible based on Vue Route docs. If I'm mistaken, then these two options are back on the table.

In the meantime, either the current solution, or the one I proposed yesterday in my previous comment, are all I can think of that would be practical. I agree with your concerns @taoeffect about this solution, but my other solution (renaming query params) is more invasive and therefore might have other ramifications on the rest of the system which, in my current lack of familiarity with the whole system, I may not yet be able to predict.

Reply:
One other solution I thought of is to use `beforeRouteLeave` on each component to manually remove its own query params. This solution is less clean, more of a DRY violation, and easier to get wrong. But it could be improved with abstraction.

The first layer of abstraction could be to create a shared function and pass the parameters to that:

```js
beforeRouteLeave() {
  removeQueryParams('section', 'filter') // alternatively, SBP could be used
}
```

This is roughly the same complexity as the current solution, and might actually be cleaner both in interface and implementation.

This could further be abstracted by some kind of "inheritance" in order to make this "automatic":

```js
inheritFrom: 'ModalThing', // pseudo-code
queryParamsToRemove: ['section', 'filter']
```

This is less clean IMO and should be avoided, but putting it out here in case it might give anyone more ideas, or maybe my judgment about it being a poor solution is generally not agreed.

Reply:
We really need a simple, elegant, automated way of doing this that makes it difficult for developers to "forget" to do it correctly. We have all of the information we need in the URL. Over on Slack you came up with an interesting method that could show promise, by using the hash: `/app/dashboard#UserSettingsModal?section=appearance`

For example, instead of a very ugly and verbose way of associating variables (e.g. `UserSettingsModal-action`), we could try something along these lines:

```
/app/dashboard?modal#Modal1?var1=val1&var2=val2_Modal2?var1=val1_Modal3
```

Or something to that effect. I am honestly not familiar with what is and isn't possible in these URL schemes, and I wonder if with enough creativity/experimentation we could figure out a solution that allows us to specify:

1. The modals and their order
2. The variables for each modal (in a non-ugly way)

And still have everything automatically cleared when the back/forward buttons are used, or when modals are closed.

If a solution is found, it shouldn't be too difficult to go through the entire codebase and update all modal usage to use this new method.

Reply:
The only issue I see is how do you then handle using `router-link` in a clean way? I'm honestly not sure.

This was the example Sandrina gave:

```pug
router-link.link(:to='{ query: { modal: "GroupCreationModal" }}') Add a group
```

But I'm not sure what to do if you want this sort of push/pop behavior for adding modals with parameters.

Reply:
Are the only variables we ever want to remove from a modal just "identifying" information?

In the PaymentDetails modal we only have `id`, in ChangeVotingRules we only have `rule`, in RemoveMember we only have `username`, in UserSettingsModal we only have `section`. So it seems to be the case.

If that's the case, we can skip query params altogether for this, and just use a dot-separated hashtag, where removing the hashtag is sufficient to remove the modal's "navigational" information, and the modal itself can parse it from the tag:

* `#PaymentDetails.123` - it can parse `{ id: 123 }`
* `#RemoveMember.fred` - it can parse `{ username: 'fred' }`
* etc

Reply:
The only problems that still remain here are that:

* The variable of the hashtag can't have a dot in it, or we have to have a foolproof way of escaping it.
* Variables will have to be unnamed and strongly-ordered, but that's probably okay.
* There's still no way to remove actual query params that the modal may use, e.g. if the modal contains a filterable list, it may have added `?filter=foo`, which we can't automatically know to remove.

Reply:
@sdegutis That looks much nicer. I feel like that is a worthwhile direction to explore. You can play with the syntax, and consider what it might look like if you need to pass in multiple parameters to a modal.

Reply:
I don't know what you mean by removing (I'm forgetting atm the details of this, just remember what I didn't like 😆).

What about: `#PaymentDetails.id=123.id2=345#RemoveMember.username=fred` or something to that effect?

You'd have to figure out whether multiple `#` are OK after the first `#`, and if not, find a different separator, etc., and same goes for `.` and `=`, different characters can be used to do the separation.

There is still the remaining item of figuring out a push-based syntax for "click this link to add a new modal and a new corresponding set a params to the URL" to replace the previous pug stuff:

```pug
router-link.link(:to='{ query: { modal: "GroupCreationModal" }}') Add a group
```

Reply:
@sdegutis Oh, and another crazy idea: you go a completely different way too: instead of trying to effectively re-create a URL syntax to encode an object-like structure, you could pass JSON directly in the URL as base64 that's encoded and decoded.

E.g. `/app/dashboard?modal=9823uadfjop8983ru9823rf98hf982h0390gh39r0ghehsdhfsdkjfidjdfghlksdjfghkasdhff982hf22983hfdsjflj` (with or without a hash).

Just another thought that popped into my head, to kinda illustrate how much freedom there is to explore design here.

Reply:
By removing, I mean removing the modal's info from the URL when the modal closes. This is the reason @sandrina-p added the current solution, so that it can track which params to remove.

The router link will get even easier:

```js
router-link.link(:to='"#RemoveMember." + username')
```

It could use the object-based API instead but that's not much cleaner/shorter, it just replaces "#..." with "{ hash: ... }":

```js
router-link.link(:to='{ hash: "RemoveMember." + username' })
```

I think modals are inherently supposed to be small, short-lived, and single-purpose, so we can probably just not worry about ordering and naming of modal-specific variables, and live with the small but reasonable caveats listed above.

Reply:
> Oh, and another crazy idea: base64 JSON directly in the URL

Crazy indeed. One of the best benefits of URLs is that they're human readable and thus self-describing. The Java communities tried many experiments in (ab)using URLs like this, and I think there's general consensus that we just shouldn't.

Reply:
> to kinda illustrate how much freedom there is to explore design here.

Absolutely, thanks for putting it out there. I do think we're on a good path with the hashtag idea, but I've found that brainstorming like this has been very helpful in the past in coming up with out-of-the-box ideas that we can merge with other ideas to have ideas that are half-in and half-out of the box and work wonderfully.